# Tutorial based on "A Little Book of Python for Multivariate Analysis"
Author: Fermín Huarte Larrañaga

Date: Jan'26

Credit to:

- "*A Little Book of Python for Multivariate Analysis*"
- Author: Yiannis Gatsoulis
- License:  Creative Commons Attribution-ShareAlike 4.0 International License
- [Link](https://python-for-multivariate-analysis.readthedocs.io/index.html).

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

/home/ramon/anaconda3/lib/python3.11/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


### Reading Data into Python

The file [wine.data](http://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data) contains data on concentrations of 13 different chemicals in wines grown in the same region in Italy that are derived from three different cultivated varieties. 

There is one row per wine sample. The first column contains the cultivated variety identifier of a wine sample (labelled 1, 2 or 3), and the following thirteen columns contain the concentrations of the 13 different chemicals in that sample. The columns are separated by commas and there is no header in the beginning of the file, this is actually the default for `pd.read_csv()`).

In [2]:
wine_data = pd.read_csv('../data/wine.data')

We add column names ''V1'', ''V2'', and so forth where ''V'' stands for *Variable*. The first column is converted to a string type.

In [3]:
wine_data.columns = ["V"+str(i) for i in range(1, len(wine_data.columns)+1)]  # rename column names to be similar to R naming convention
wine_data.V1 = wine_data.V1.astype(str)
wine_data

,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14
0,1,13.20,1.78,2.14,11.2,100,2.65,2.76,0.26,1.28,4.38,1.05,3.40,1050
1,1,13.16,2.36,2.67,18.6,101,2.80,3.24,0.30,2.81,5.68,1.03,3.17,1185
2,1,14.37,1.95,2.50,16.8,113,3.85,3.49,0.24,2.18,7.80,0.86,3.45,1480
3,1,13.24,2.59,2.87,21.0,118,2.80,2.69,0.39,1.82,4.32,1.04,2.93,735
4,1,14.20,1.76,2.45,15.2,112,3.27,3.39,0.34,1.97,6.75,1.05,2.85,1450
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
172,3,13.71,5.65,2.45,20.5,95,1.68,0.61,0.52,1.06,7.70,0.64,1.74,740
173,3,13.40,3.91,2.48,23.0,102,1.80,0.75,0.43,1.41,7.30,0.70,1.56,750
174,3,13.27,4.28,2.26,20.0,120,1.59,0.69,0.43,1.35,10.20,0.59,1.56,835
175,3,13.17,2.59,2.37,20.0,120,1.65,0.68,0.53,1.46,9.30,0.60,1.62,840


Now, **V1** contains the **dependent variable data**, that is the wine variety. This value (1, 2, 3) is attributed according to the values of the following **independent variable data**, **V2** to **V14**.


In [12]:
# independent variables
X = wine_data.loc[:, "V2":]
# dependent variable
y = wine_data.V1

**Exercise:** How many wine samples have been collected? (*Answer:117*) How many for each wine variety? (*Answer:58 samples of type 1, 71 of type 2, 48 of type 3*)

In [ ]:
len(y[y=='3'])

In [ ]:
wine_data.groupby(['V1']).count().V2

### Plotting Multivariate Data
Once you have read a multivariate data set into python, the next step is usually to make a plot of the data.

#### Matrix Scatterplot

One common way of plotting multivariate data is to make a matrix scatterplot, showing each pair of variables plotted against each other. This can allow for a quick scan of correlation between variables in the data set. We can use the `scatter_matrix()` function.

In order to do that, we need to us as the function's input the variables that we want to include in the plot. Let us start by just including the variables corresponding to the concentrations of **the first five chemicals**, that is columns V2-V6 of the variable data. We can extract just these columns from the variable data by typing:

In [ ]:
wine_data.loc[:,'V2':'V6']

Then we use the `scatter_matrix()` function to make the scatter matrix plot of these variables:



In [ ]:
pd.plotting.scatter_matrix(wine_data.loc[:, 'V2':'V11'], diagonal='hist');

Each of the off-diagonal cells is a scatterplot of two of the five chemicals, for example, the second cell in the first row is a scatterplot of V2 (y-axis) against V3 (x-axis).

Notice that the plots in the diagonal cells are Vi-Vi autocorrelations. The parameter `diagonal` in the function allows us to specify whether to plot a histogram (`hist`) or a Kernel Density Estimation (`kde`) for the variable. 

In [ ]:
pd.plotting.scatter_matrix(wine_data.loc[:, 'V2':'V6'], diagonal='kde');

If you see an interesting scatterplot for two variables in a matrix scatterplot, you may want to plot that actual scatterplot in more detail distinguishing the data according to the value of the dependent variable (their wine variety in this case).

For example, in the matrix scatterplot above, the cell in the third column of the fourth row down is a scatterplot of V5 (x-axis) against V4 (y-axis). If you look at this scatterplot, it appears that there may be a positive correlation between V5 and V4.



In [ ]:
wine_data.plot.scatter('V4','V5')

One may therefore decide to examine the relationship between V5 and V4 more closely, by plotting a scatterplot of these two variables, with the data points coloured according to the corresponding wine variety. A fast way to do this is provided by the powerful module `seaborn` linked to `pandas`.

In [ ]:
sns.scatterplot(data=wine_data, x='V4', y='V5', hue='V1')

In [ ]:
sns.scatterplot(data=wine_data, x='V4', y='V8', hue='V1')

The scatterplot shows othat the wines from variety 2 seem to have lower values of V4 compared to the wines of variety 1.

#### A Profile plot

Another type of plot that is useful is a `profile plot`, which shows the **variation in each of the variables**, by plotting the value of each of the variables for each of the samples.

In [ ]:
wine_plot = wine_data.loc[:, 'V2':'V6'].plot()
wine_plot.legend(loc='center right')

It is clear from the profile plot that the mean and standard deviation for V6 is quite a lot higher than that for the other variables.

### Calculating Summary Statistics for Multivariate Data
Calculate summary statistics such as the mean and standard deviation for each of the variables in the multivariate data set. This is easy to do, using the `mean()` and `std()` functions in `numpy` and applying them to the Pandas dataframe.

Pandas allows to do simple operations directly calling them as methods, for example we could do compute the means of a dataframe `df` by calling `df.mean()`. Notice that it makes no sense to compute statistics of the dependent variable ''y'', V1 in our case.

Remember some cells above we have included all the dependent variables to ''X'':

In [5]:
X.mean()

V2      12.993672
V3       2.339887
V4       2.366158
V5      19.516949
V6      99.587571
V7       2.292260
V8       2.023446
V9       0.362316
V10      1.586949
V11      5.054802
V12      0.956983
V13      2.604294
V14    745.096045
dtype: float64

An alternative option is to use the `apply` method of the `pandas.DataFrame` class, which applies the passed argument function along the input axis of the DataFrame. *This method is powerful as it allows passing any function we want to be applied in our data*.

So we can equivalently type:

In [6]:
X.apply(np.mean)

V2      12.993672
V3       2.339887
V4       2.366158
V5      19.516949
V6      99.587571
V7       2.292260
V8       2.023446
V9       0.362316
V10      1.586949
V11      5.054802
V12      0.956983
V13      2.604294
V14    745.096045
dtype: float64

Notice the species concentrations have quite different values:
- V14 and V6 appear to be very large, 
- V2 and V5 are about one order of magnitude smaller,
- the rest of the concentrations are smaller than 10

Let us have a look at the standard deviations:

**Exercise XX:** Compute the standard deviation for the concentration of each chemical species in the data set.

In [7]:
X.apply(np.std)

V2       0.806520
V3       1.116148
V4       0.274302
V5       3.326634
V6      14.133922
V7       0.624693
V8       0.995833
V9       0.124300
V10      0.569928
V11      2.317871
V12      0.228487
V13      0.703108
V14    313.993283
dtype: float64

Again, the standard deviation value for V14 is significantly larger than the rest, also that of V6 and V11 is quite large. It would therefore make sense to standardise in order to compare the variables so that each variable has a sample variance of 1 and sample mean of 0.

#### Means and Variances per Group
It can be interesting to calculate the means and standard deviations for just the samples from a particular group, for example, for the wine samples from each variety. The wine variety is stored in the V1 column of the variable data, which we have previously assigned to ''y'' for convenience.

Let us select the data only for those sample where the wine variety is ''2'', this means, column V1 contains a 2. Important! Keep in mind that we defined V1 values as `string`!!

In [ ]:
variety_2_data = wine_data[y == '2']

Equivalently

In [ ]:
variety_2_data = wine_data[wine_data.V1 == '2']

We can now calculate the mean and standard deviations of the 13 chemicals’ concentrations, for just the samples of the ''2'' wine variety:

In [ ]:
variety_2_data.loc[:, "V2":].apply(np.mean)

In [ ]:
variety_2_data.loc[:, "V2":].apply(np.std)

Do it all at once, for each variety

In [ ]:
X.groupby(y).apply(np.mean, axis=0)

In [ ]:
X.groupby(y).apply(np.std, axis=0)

Number of samples, for each variety

In [ ]:
X.groupby(y).apply(len)

#### Between-group Variance and Within-group Variance for a Variable

- [Within-group variance](https://www.statisticshowto.com/within-group-variation/#:~:text=Within%2DGroup%20Variation:%20Definition%20and,SSW:%20Sum%20of%20Squares%20Within.) measures the spread or variability of individual data points inside each group, representing random error or natural differences not explained by the group differences. It represents "noise" or random fluctuations not caused by the independent variables being tested.

- [Between-group variance](https://www.statisticshowto.com/between-group-variation) measures how much the means of different groups vary from the overall mean, indicating differences attributable to the grouping factor (like a treatment or condition). It's the "*explained*" variation, showing if group averages are truly different from each other, contrasting with within-group variance (random error). 

A large between-group variance relative to within-group variance suggests the factor distinguishing the groups has a significant effect. 

**Calculate within-group variance:**

$$ \frac{\sum_{i=1}^g \sum_{j=1}^{g_{samp}} \left ( X_{ij} - {\overline{X_i}} \right )^2}{N - g}$$
 
where $g$ is the number of groups, $N$ the total number of data points,  $g_{samp}$ is the number of data points belonging to each group, $X_{ij}$ is the value of an individual data point belonging to group $i$, ${\overline{X_i}}$ is the mean value of X in group $i$

If we want to calculate the within-groups variance for a particular variable (for example, ''V2''), we can use the function below:

In [ ]:
groups = sorted(set(y)) # a python set withe unique values in y (i.e. V1 column)
ngroups = len(groups);

In [ ]:
#Compute variance within the group
numtotal = 0
denomtotal = 0
for g in groups:
    g_data = X.V2[y == g]
    l = len(g_data)
    #variance of the elements belonging to the same group
    var_g = np.var(g_data) #g_data.var() Population variance vs Sample Variance
    numg = l*var_g #weighted by the number of samples
    denomg= l
    numtotal = numtotal + numg
    denomtotal = denomtotal + denomg
Vw = numtotal / (denomtotal - ngroups)
Vw

**Calculate between-group variance:**

$$ \frac{\sum_{i=1}^g g_{samp} \left (  {\overline{X}} - {\overline{X_i}} \right )^2}{g - 1}$$

If we want to calculate the within-groups variance for a particular variable (for example, ''V2''), we can use the function below:

In [ ]:
ss = []
mean_overall = np.mean(X.V2) #overall mean
for g in groups:
    g_data = X.V2[y == g]
    l = len(g_data)
    #mean value in the group
    mean_g = np.mean(g_data)
    #variation to the overall mean
    numg = (mean_overall - mean_g)**2 
    ss.append(l*numg) #weighted by the number of samples
# calculate the between-groups variance
Vb = sum(ss) / (ngroups - 1)
Vb

We can calculate the **separation achieved by the V2 variable** as its between-groups variance divided by its within-groups variance. Thus, the separation achieved by V2 is calculated as:

In [ ]:
Vb/Vw

In order to be more sistematic, we can shift the previous calculations of both variances to user-defined functions:

In [ ]:
def calc_var_within_group(variable, groups):
    #Compute variance within the group
    numtotal = 0
    denomtotal = 0
    for g in groups:
        g_data = variable[y == g]
        l = len(g_data)
        #variance of the elements belonging to the same group
        var_g = np.var(g_data) #g_data.var() Population variance vs Sample Variance
        numg = l*var_g #weighted by the number of samples
        denomg= l
        numtotal = numtotal + numg
        denomtotal = denomtotal + denomg
    Vw = numtotal / (denomtotal - ngroups)
    return Vw

In [ ]:
def calc_var_between_group(variable, groups):
    #Compute variance between the groups
    ss = []
    mean_overall = np.mean(variable) #overall mean
    for g in groups:
        g_data = variable[y == g]
        l = len(g_data)
        #mean value in the group
        mean_g = np.mean(g_data)
        #variation to the overall mean
        numg = (mean_overall - mean_g)**2 
        ss.append(l*numg) #weighted by the number of samples
    # calculate the between-groups variance
    Vb = sum(ss) / (ngroups - 1)
    return Vb

In [ ]:
calc_var_within_group(X.V2,groups)

In [ ]:
calc_var_between_group(X.V2, groups)

The *separation* introduced by V2 in the group is calculated as the ratio of both variances:

$$ sep = \frac{Vb}{Vw} $$

In [ ]:
calc_var_between_group(X.V2, groups)/calc_var_within_group(X.V2,groups)

If we want now to calculate the separations achieved by all of the variables in a multivariate data set, we can run the both functions for each independent variable:

In [ ]:
for V in X: #for each column in the independent variables data set
        Vw = calc_var_within_group(X[V],groups)
        Vb = calc_var_between_group(X[V], groups)
        sep = Vb/Vw
        print(f"var: {V:3s} Vw= {Vw:10.4f} \t Vb={Vb:10.4f} \t separation={sep:10.4f}")

Let's have a look at the table. **The individual ''V'' variable which gives the greatest separations between the groups (the wine varieties) is ''V8'' (separation 230.7).** However, keep in mind that this is a *multidimensional data set*. While V8 is the best single variable to discriminate data, there may be a linear combination of the individual variables that might give a larger separation between the groups.

#### Between-groups Covariance and Within-groups Covariance for Two Variables

See notes on covariance:

$$\sigma_{xy} = \sum_{j=1}^N (x_j - {\overline{x}})(y_j - {\overline{y}})$$

In our case, we have within-group covariance

$$\sigma_{xy} = \frac{\sum_{i=1}^g \sum_{j=1}^{g_{samp}} (x_{ij} - {\overline{x_i}})(y_{ij} - {\overline{y_i}})}{N - g}$$

In [ ]:
def Covariance_within_Group(var1, var2, groups):
    Covw = 0.0
    for g in groups:
        cc = []
        g_data1 =  var1[ y == g ]
        g_data2 =  var2[ y == g ]
        g_mean1 = np.mean(g_data1)
        g_mean2 = np.mean(g_data2)
        cc = [(d1 - g_mean1)*(d2 - g_mean2) for d1,d2 in zip(g_data1, g_data2)]
        cov_group = sum(cc)
        Covw = Covw + cov_group
    Covw = Covw / (len(var1) - ngroups)
    return Covw                        

Covariance within group for variables V8 and V11:

In [ ]:
Covariance_within_Group(X.V8, X.V11, groups)

In the case of the between-groups covariance:


$$ \frac{\sum_{i=1}^g g_{samp} \left (  {\overline{X}} - {\overline{X_i}} \right ) \left (  {\overline{Y}} - {\overline{Y_i}} \right )}{g - 1}$$

In [ ]:
def  Covariance_between_Groups(var1, var2, groups):
    Covb = []
    #calculate overall means
    mean1 = np.mean(var1)
    mean2 = np.mean(var2)
    #between groups covariance
    for g in groups:
        g_data1 =  var1[ y == g ]
        g_data2 =  var2[ y == g ]
        g_mean1 = np.mean(g_data1)
        g_mean2 = np.mean(g_data2)
        Covb.append((g_mean1 - mean1)*(g_mean2 - mean2)*len(g_data1))
    Covb = sum(Covb) / (ngroups - 1)
    return Covb                        

In [ ]:
Covariance_between_Groups(X.V8, X.V11, groups)

Thus, for V8 and V11, the between-groups covariance is -60.71 and the within-groups covariance is 0.29. Since the within-groups covariance is positive (0.29), it means V8 and V11 are positively related within groups: for individual data points from the same group, individuals with a high value of V8 tend to have a high value of V11, and vice versa. Since the between-groups covariance is negative (-60.71), V8 and V11 are negatively related between groups: groups with a high mean value of V8 tend to have a low mean value of V11, and vice versa.

#### Calculating Correlations for Multivariate Data

It is often interesting to investigate whether any of the variables in a multivariate data set are significantly correlated. To calculate the linear (Pearson) correlation coefficient ([PCC](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient)) between all the variables in the data set we can used the `DataFrame.corr()` method from Pandas and get the **correlation matrix**:

In [ ]:
corr_matrix = X.corr()
corr_matrix

The matrix can be rapidly inspected and we can identify larger elements such as V8/V7 (0.79) or V11/V12 (-0.52, anticorrelation). A nice way to do it is graphically using the tools provided by Seaborn:

In [ ]:
sns.heatmap(corr_matrix, vmax=1., square=False).xaxis.tick_top()

A very elegant alternative visualization is via a [Hinton diagram](https://matplotlib.org/stable/gallery/specialty_plots/hinton_demo.html). The color of the boxes determines the sign of the correlation, in this case red for positive and blue for negative correlations; while the size of the boxes determines their magnitude, the bigger the box the higher the magnitude.

In [ ]:
# adapted from http://matplotlib.org/examples/specialty_plots/hinton_demo.html
# adapted from https://python-for-multivariate-analysis.readthedocs.io/a_little_book_of_python_for_multivariate_analysis.html#between-groups-covariance-and-within-groups-covariance-for-two-variables
def hinton(matrix, max_weight=None, ax=None):
    """Draw Hinton diagram for visualizing a correlation matrix."""
    ax = ax if ax is not None else plt.gca()

    if not max_weight:
        max_weight = 2**np.ceil(np.log(np.abs(matrix).max())/np.log(2))

    ax.patch.set_facecolor('lightgray')
    ax.set_aspect('equal', 'box')
    ax.xaxis.set_major_locator(plt.NullLocator())
    ax.yaxis.set_major_locator(plt.NullLocator())

    for (x, y), w in np.ndenumerate(matrix):
        color = 'blue' if w > 0 else 'red'
        size = np.sqrt(np.abs(w))
        rect = plt.Rectangle([x - size / 2, y - size / 2], size, size,
                             facecolor=color, edgecolor=color)
        ax.add_patch(rect)

    nticks = matrix.shape[0]
    ax.xaxis.tick_top()
    ax.set_xticks(range(nticks))
    ax.set_xticklabels(list(matrix.columns), rotation=90)
    ax.set_yticks(range(nticks))
    ax.set_yticklabels(matrix.columns)
    ax.grid(False)

    ax.autoscale_view()
    ax.invert_yaxis()

hinton(corr_matrix)

Large blue squares identify variables that correlate strongly and positively while large red squares are found the air of variables have a strong negative correlation.

Although the correlation matrix and diagrams are useful for quickly looking to identify the strongest correlations, they still some work to find the n-strongest correlations. 
In order to this we are goign to write a short piece of code:

In [ ]:
mask = np.triu(np.ones(corr_matrix.shape),k=1).astype(bool) #mask to delete all the data from the diagonal and lower triangle
corr_masked = corr_matrix.where(mask) #apply the mask
corr_masked = corr_masked.stack() #stack the data using two indeces (row and column)
corr_masked = corr_masked.abs().sort_values(ascending=False).reset_index() #sort the values and reset index
# 10 highest values of the correlation matrix
top=10
display(corr_masked.head(top))

This tells us that the pair of variables with the highest linear correlation coefficient are V7 and V8 (correlation = 0.86 approximately). We can pllot the correspoding scatterplot:

In [ ]:
sns.scatterplot(data=wine_data, x='V7', y='V8', hue='V1')

### Standardising Variables

If we want to compare different variables that have different units, and therefore very different variances, it is a good idea to **first standardise the variables**.
In our example, we found above that the concentrations of the 13 chemicals in the wine samples show a wide range of standard deviations, from 0.124 for V9 (variance 0.015) to 314.022 for V14 (variance 98609.60). In this case, it is not a good idea to use the unstandardised chemical concentrations as the input for a principal component analysis (PCA, see below) of the wine samples, as if we did that, the first principal component would be dominated by the variables which show the largest variances, such as V14.

Therefore, it is highhly advisable to first standardise the variables so that **they all have variance 1 and mean 0**, and to then carry out the principal component analysis on the standardised data. This would allow us to find the principal components that provide the best low-dimensional representation of the variation in the original data, without being overly biased by those variables that show the most variance in the original data.



#### Standardization using Python

Standardization, also known as Z-score normalization, involves transforming the data such that it has a mean of zero and a standard deviation of one. Mathematically, the standardization formula for a feature 𝑥 is given by:

$$ z = \frac{x - \mu}{\sigma}$$

where
- 𝑧 is the standardized value,
- 𝑥 is the original value,
- $\mu$ is the mean and,
- $\sigma$ is the standard deviation.

In [8]:
from sklearn.preprocessing import scale

In [20]:
X.dtypes

V2     float64
V3     float64
V4     float64
V5     float64
V6       int64
V7     float64
V8     float64
V9     float64
V10    float64
V11    float64
V12    float64
V13    float64
V14      int64
dtype: object

SciKitLearn module can help us simplify this (not so hard) task by using the `scale()` function from the package `sklearn.preprocessing`.

For example, to standardise the concentrations of the 13 chemicals in the wine samples, we type:

In [28]:
X_std = X.copy().astype(np.float64)
X_std.loc[:,:] = scale(X)
X_std

,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14
0,0.255824,-0.501624,-0.824485,-2.500110,0.029180,0.572666,0.739636,-0.823139,-0.538575,-0.291130,0.407100,1.131698,0.971052
1,0.206229,0.018020,1.107690,-0.275639,0.099932,0.812784,1.221645,-0.501337,2.145975,0.269729,0.319567,0.804579,1.400998
2,1.706501,-0.349315,0.487935,-0.816726,0.948953,2.493609,1.472691,-0.984039,1.040572,1.184362,-0.424458,1.202811,2.340509
3,0.305420,0.224086,1.836812,0.445811,1.302712,0.812784,0.669343,0.222716,0.408913,-0.317016,0.363334,0.463238,-0.032154
4,1.495719,-0.519543,0.305655,-1.297693,0.878201,1.565153,1.372273,-0.179536,0.672104,0.731360,0.407100,0.349457,2.244965
...,...,...,...,...,...,...,...,...,...,...,...,...,...
172,0.888171,2.965658,0.305655,0.295509,-0.324579,-0.980097,-1.419361,1.268570,-0.924589,1.141219,-1.387314,-1.229247,-0.016230
173,0.503803,1.406725,0.415023,1.047020,0.170684,-0.788003,-1.278776,0.544517,-0.310476,0.968647,-1.124717,-1.485253,0.015618
174,0.342617,1.738222,-0.387012,0.145207,1.444215,-1.124168,-1.339027,0.544517,-0.415753,2.219795,-1.606145,-1.485253,0.286324
175,0.218628,0.224086,0.014006,0.145207,1.444215,-1.028121,-1.349068,1.349020,-0.222746,1.831507,-1.562379,-1.399918,0.302248


In [27]:
y.astype(np.int32)

0      1
1      1
2      1
3      1
4      1
      ..
172    3
173    3
174    3
175    3
176    3
Name: V1, Length: 177, dtype: int32

For the sake of clarity, let us call these standardised variables ''Z2, Z3, Z4, ... Z14'' and rename the columns of the data frame accordingly:

**Exercise:** Compute the mean and variance for each column in standardised X. Check that mean is close to zero and variance is one.

## Principal Component Analysis

PCA (Principal Component Analysis) is a dimensionality reduction technique and helps us to reduce the number of features in a dataset while keeping the most important information. It changes complex datasets by transforming correlated features into a smaller set of uncorrelated components. 

PCA uses linear algebra to transform data into new features called principal components. It finds these by calculating eigenvectors (directions) and eigenvalues (importance) from the covariance matrix. PCA selects the top components with the highest eigenvalues and projects the data onto them simplify the dataset. It prioritizes the directions where the data varies the most because more variation means more useful information.

In our example, we have 13 chemical concentrations describing wine samples from three different varieties. We can carry out PCA to see if we can get most of the variation between samples using a smaller set of variables (these will be our *principal components*). These new variables will be a linear combination of all or some of the 13 chemical concentrations.

Typically, the first step is often to **standardise the variables** under study function. We just did that in the previous section (see above). 

Once we have standardised our variables, we can carry out a principal component analysis using the **[`PCA` class](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn.decomposition.PCA)** from `sklearn.decomposition`  package and its [`fit` method](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn.decomposition.PCA.fit). This method will fit the model with our standardised data stored in variable `standardised_X`. The default solver is [**Singular Value Decomposition**](https://en.wikipedia.org/wiki/Singular_value_decomposition) (“svd”). Keep in mind that `PCA().fit()` calculates the necessary parameters (eigenvectors and eigenvalues) to define the new space storing them within the fitted PCA object, but **it does not return the transformed dataset**. For more information you can type help(PCA) in the python console.

Let's thus carry out a principal components analysis on the standardised concentrations by first import the `PCA class`:

and then typing

by the default, the number of components will be the same as the number of standardised variables. We could feed the PCA class an optional argument `n_components=` and limit the number of components.

Now `pca` is an instance to the `PCA` class and its main attributes are:
- `pca.explained_variance_`: a 1D Numpy (column) array containing the amount of variance explained by each of the selected components. These largest eigenvalues of the covariance matrix of X. In our case, 13.
- `pca.explained_variance_ratio`: a 1D Numpy (column) array containing the ratio of variance explained by each of the components. The sum of the elements in this array is 1.
- `pca.components_`: a Numpy array containing the directions of maximum variance in the data. In our case it is $13\times13$

So the component with the largest variance corresponds to the first (and largest) eigenvalue in `pca.explained_variance_`

and this component ammounts for

We can list the principal components (named PC1, PC2, and so on) according to the magnitude of the variance:

Notice that the first three principal components already explain almost 70% of the variance in the data. This largest principal component, which we will name **PC1** is a linear combination of the standardised variables "Z" and the linear combination is given by `pca.explained_variance_`. PC1 is given by the first row:

In a (bit) nicer form:

These linear combination coefficients are referred to as *loadings* in the context of PCA. The loadings quantify the relationship between the original variables and the newly created principal components (PCs). Ranging from -1 to +1, they indicate how much each variable contributes to a PC, acting as weights that define the direction and strength of that relationship. PC1 has highest (in absolute value) loadings for V8, V7, V13, V10, V12, V14, V9, V3, and V5. The loadings for V8, V7, V13, V10, V12 and V14 are of the same sign, while those for V9, V3, and V5 are of opposite sign. Therefore, an interpretation of the PC1 is that it represents a contrast between the concentrations of V8, V7, V13, V10, V12, and V14, and the concentrations of V9, V3 and V5


Note that sum of the square of the coefficients is one:

#### Deciding how many Principal Components to retain
So far we have reduced nothing, we had 13 variables (V2 to V14) in the original data, we have built 13 standardised variable (Z2 to Z14) and the PCA procedure has yielded 13 principal components (PC1 to PC13). However, we just saw above that we have now a hierarchy and some components explain most of the variance in our data. This allows us to reduce the dimensionality of our analysis. In order to decide how many principal components should be retained, it is common to summarise the results of a principal components analysis by making a [scree plot](https://en.wikipedia.org/wiki/Scree_plot). This consists of a simple line plot of the variance values:

In [ ]:
yy = pca.explained_variance_
x = np.arange(1,len(pca.explained_variance_)+1)
plt.plot(x, yy, "o-")
plt.xticks(x, pc_labels, rotation=60)
plt.ylabel("Variance")
plt.grid()
plt.show()

According to the literature several criteria can be applied to selected the reduced set of PCs to proceed with the analysis:
- One possible criterion is to observe that a change in slope in the scree plot occurs at component 4. Therefore, it could be argued that the first three components should be retained.
- Another way of deciding how many components to retain is to use *Kaiser’s criterion*: only retain principal components for which the variance is above 1 (standardised data!!). We would retain, again, the first three principal components.
- A third way to decide how many PCs to retain is to decide to keep the number of components required to explain at least some minimum amount of the total variance. For example, to explain at least 80% of the variance, we would retain the first five principal components, as we can see from the table we computed above.



As mentioned before, `PCA().fit()` calculates the transformation (eigenvectors and eigenvalues) to the new space but **it does not return the transformed dataset**. In order to get that we use the `PCA().transform()` method:

`transformed_data` is a new (177,14) Pandas DataFrame where the first column contains the wine variety label and the remaining 13 columns contain the data of the corresponding 13 PCs for each of the 177 samples. So  the values of the first principal component, PC1, for each sample in our wine data are obtained as:

#### Scatterplots of the Principal Components

We can plot the transformed data. Let us build scatterplot of the first two principal components, and label the data points according to the wine variety of the sample (remember this is V1 and was stored in y):

Again, a nicer alternative using [`lmplot` from `Seaborn`](https://seaborn.pydata.org/generated/seaborn.lmplot.html):

The scatterplot shows the first principal component on the x-axis, and the second principal component on the y-axis. We can see from the scatterplot that wine samples of variety 3 have clearly lower values of PC1 than wine samples of variety 3. **Therefore, PC1 can be used to separate wine samples of variety 1 from those of variety 3.** We can also see that wine samples of variety 2 have lower values of the second principal component than wine samples of varieties 1 and 3. **Therefore, PC2 can be used to separate samples of group 2 from samples of groups 1 and 3.**

We can thus conclude that PC1 and PC2 are reasonably useful for distinguishing wine samples of the three different varieties.

## Linear Discriminant Analysis
The purpose of principal component analysis is to find the best low-dimensional representation of the variation in a multivariate data set. In our wine data set example, we have 13 chemical concentrations describing wine samples from three wine varieties. By carrying out a PCA, we have found that most of the variation in the chemical concentrations between the samples can be captured using the first two principal components, PC1 and PC2. Each of the principal components is a particular linear combination of the 13 chemical concentrations.

The purpose of linear discriminant analysis (LDA) is to find the linear combinations of the original variables (the 13 chemical concentrations here) that gives the best possible separation between the groups in the data set. Notice the difference with PCA where we also seek linear combinations. In LDA we aim for the best group separation, we rely on having previously labelled the data points in groups. In PCA the linear combination seeks the largest contributions to the variance without any knowledge of the labelling.

The wines in our data set come from three different varieties, so the number of groups (G) is 3, and the number of variables is 13 (13 chemicals’ concentrations; p = 13). The maximum number of useful discriminant functions that can separate the wines by cultivar is the minimum of G-1 and p, and so in this case it is the minimum of 2 and 13, which is 2. Thus, we can find at most 2 useful discriminant functions to separate the wines by cultivar, using the 13 chemical concentration variables.

We will carry out a linear discriminant analysis by using the `LinearDiscriminantAnalysis` class model from the module `sklearn.discriminant_analysis` and using its method `fit()` to fit our X, y data.

In [ ]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

To carry out a linear discriminant analysis using the 13 chemical concentrations in the wine samples, we type:

Notice that while in the case of the `PCA().fit()` we only provided the (independent) variable values, here we are giving X and also y, the dependent variable values. The values of the coefficients of the discriminant functions for the wine data are stored in the `scalings_` attribute of the `lda` object model.`lda.scalings_` is a NumPy array of 13 rows and 2 columns that contains the linear combination for the two linear discriminant variables (LD1 and LD2) in terms of the original V2 to V14 variables. Let's print them

The values of the discriminant functions can be calculated using  the `transform(X)` or `fit_transform(X, y)` methods of the `LDA` object:

`lda_transformed_data` is a new (177,3) Pandas DataFrame where the first column contains the wine variety label and the two remaining columns contain the values of LD1 and LD2 for each of the 177 samples, respectively. So  the values of the linear discriminant function, LD1, for each sample in our wine data are obtained as:

*Note:* It doesn’t matter whether the input variables for linear discriminant analysis are standardised or not, unlike for PCA. However, using standardised variables in LDA makes it easier to interpret the loadings in a linear discriminant function.

Similarly to what we did in PCA we can obtain a scatterplot of the best two discriminant functions, with the data points labelled by wine variety, by typing:

Notice we get a very nice separation of the three groups. The first discriminant function (LD1, x-axis) separates varieties 1 and 3 very well, but doesn’t not perfectly separate 1 and 3, or  2 and 3. The second discriminant function (PC2, y-axis) achieves a fairly good separation of varieties 1 and 3, and varieties 2 and 3, although it is not totally perfect. To achieve a very good separation of the three varieties, it would be best to use both the first and second discriminant functions together.

### Decission rules and Missclassification rate

In LDA, decision rules (also called allocation rules or classification rules) specify how a **new observation** is assigned to one of the predefined classes after the discriminant functions have been constructed. In our example, it would mean how would we assign a new wine sample (with 13 chemical species concentrations) to a given wine variety: 1, 2, or 3. In the context of LDA, this procedure has two main steps:

- Learning step. We have already completed this step. From labeled data, LDA finds one or more discriminant functions that best separate the classes. We have LD1 and LD2.

- Decision step. Given a new observation, the allocation rule decides which class it belongs to.


The basic idea is that once once data are projected onto the LDA space, each class occupies a region. The allocation rule assigns a new point to the class for which it is “most compatible”. This compatibility can be defined in several ways:
- maximum discriminant score
- minimum distance to class mean
- maximum posterior probability



We can calculate the mean values of the discriminant functions for each of the three wine varieties:

and standard deviation:

Looking at the numbers we find that the mean value of LD1 is -3.422 for variety 1, -0.0797 for variety 2, and 4.3245 for group 3. The mid-way point between the mean values for groups 1 and 2 is -1.751, and the mid-way point between the mean values for groups 2 and 3 is 2.122.

We can therefore use the following **decision rule**:
- if $LD_1 \le -1.751$ $\Rightarrow$ predict the wine sample to be from variety 1
- if $2.122 \ge LD_1 \gt -1.751$ $\Rightarrow$  predict the sample to be from variety 2
- if $LD_1 \gt 2.122$ $\Rightarrow$  predict the sample to be from variety 3

We can check the reliability of this criterion using our wine data set:

In [ ]:
conf_m = np.zeros((3,3))
cutoff=[-1.751, 2.122]
for i,g in enumerate(groups):
    g_data = lda_transformed_data[lda_transformed_data.Group==g]
    g1 = (g_data.LD1 < cutoff[0]).sum()
    conf_m[i,0] = g1
    print(f"Number of samples from group {g} predicted as group 1 is: {g1}")
    g2 = ((cutoff[1] > g_data.LD1) & (g_data.LD1 >= cutoff[0])).sum()
    print(f"Number of samples from group {g} predicted as group 2 is: {g2}")
    conf_m[i,1] = g2
    g3 = (g_data.LD1 >= cutoff[1]).sum()
    print(f"Number of samples from group {g} predicted as group 3 is: {g3}")
    conf_m[i,2] = g3
    print("===")

This information can be displayed in what is called a [*confusion matrix*](https://en.wikipedia.org/wiki/Confusion_matrix):

In [ ]:
print("Confusion Matrix:")
print(conf_m)

There are 3+6+1=10 wine samples that are misclassified, out of 177 wine samples. According to our LDA and prediction criteria 3 samples from variety 1 are assigned to group 2, 6 samples from variety 2 are to group 1, and 1 is predicted to be from group 3. All the samples from variety 3 are classified correctly. Therefore, the misclassification rate is 5.6% (10 out of 177). This a rather low value, and therefore the allocation rule appears to be relatively realible. However, this is probably an underestimate of the misclassification rate, given that we have used the *training set* as the *testing set*. If we calculated the misclassification rate for a separate test set consisting of data other than that used to make the allocation rule, we would probably get a higher estimate of the misclassification rate.

The [**scikit-learn** Python module](https://scikit-learn.org/stable/) allows to do all the work above in a much faster way and providing extended automatic report capabilities by using the `sklearn.metrics` module. The above confusion matrix and other reporting typical performance metrics, such as:
- precision
- recall
- f1-score